cmd/snap/cmd_download.go

```diff
@@ -170,3 +163,160 @@ func (x *cmdDownload) Execute(args []string) error {
 
 	return nil
 }
+
+func (x *cmdDownload) downloadViaSnapd(snapName string, rev snap.Revision) error {
+	opts := &client.SnapOptions{
+		Channel:   x.Channel,
+		CohortKey: x.CohortKey,
+		Revision:  rev.String(),
+	}
+	fname, dlSize, stream, err := x.client.Download(snapName, opts)
+	if err != nil {
+		return err
+	}
+	fmt.Fprintf(Stdout, i18n.G("Fetching snap %q\n"), snapName)
+
+	if x.Basename != "" {
+		fname = x.Basename + ".snap"
+	}
+	downloadPath := filepath.Join(x.TargetDir, fname)
+	if err := os.MkdirAll(filepath.Dir(downloadPath), 0755); err != nil {
+		return err
+	}
+	f, err := os.Create(downloadPath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	// ensure we have progress
+	pbar := progress.MakeProgressBar()
+	pbar.Start(snapName, float64(dlSize))
+	defer pbar.Finished()
+
+	// Intercept sigint
+	c := make(chan os.Signal, 3)
+	signal.Notify(c, syscall.SIGINT)
+	go func() {
+		<-c
+		pbar.Finished()
+		stream.Close()
+	}()
+	defer signal.Reset(syscall.SIGINT)
+
+	h := crypto.SHA3_384.New()
+	mw := io.MultiWriter(f, h, pbar)
+	if _, err := io.Copy(mw, stream); err != nil {
+		return err
+	}
+	pbar.Finished()
+
+	fmt.Fprintf(Stdout, i18n.G("Fetching assertions for %q\n"), snapName)
+
+	assertFname := strings.TrimSuffix(downloadPath, filepath.Ext(downloadPath)) + ".assert"
+	sha3_384, err := asserts.EncodeDigest(crypto.SHA3_384, h.Sum(nil))
```

Comment:
can you add a `// TODO` to compare this hash with the one the store provided us?

Reply:
The code now checks the hash.